View for node and edge projections

uditagar opened the pull request:
This is the functionality to support node/edge projections.

roshandathathri reviewed (approved):
LGTM

uditagar committed:
draft PR for projection

amberhassaan commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,159 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+      const PropertyGraph* pg, const std::vector<std::string>& node_properties,
+      const std::vector<std::string>& edge_properties);
+
+  static std::unique_ptr<ProjectedShuffleTopology> Make(
```
Let's keep `Make` and delete `MakeProjectedGraph`

roshandathathri replied:
If you decide to drop `Type` from the class name (based on my previous comment), then rename this to `MakeTypeProjectedTopology`.

Also, please add documentation. Specifically, node_types means you're projecting nodes that have all the given types, right? That should be in the doxygen comments.

amberhassaan commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,182 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+std::unique_ptr<katana::ProjectedShuffleTopology>
+katana::ProjectedShuffleTopology::MakeProjectedGraph(
+    const katana::PropertyGraph* pg,
+    const std::vector<std::string>& node_properties,
```
This interface is not clear (and may be wrong). Below you're calling `GetNodeEntityTypeID(name)`, but that method takes the name of an atomic (singular) type, not a property. 
Further, composite types don't have names. Only atomic types do. 
Let's think more about this, but for now, you should take a vector of EntityTypeIDs, or a `StandardRange`.

uditagar replied:
yes, you are right. I am taking a vector of atomic node and edge types. Then I find their entity type ids inside the function

uditagar replied:
I will rename it to node_types and edge_types then

amberhassaan commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,182 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+
+    katana::do_all(katana::iterate(topology.all_nodes()), [&](auto src) {
+      for (auto type : node_entity_type_ids) {
+        if (pg->DoesNodeHaveType(src, type)) {
```
This also may be wrong, because if see the implementation of `DoesNodeHaveType`, it returns true if the type you provide is a subtype of the actual node type. You should be calling `GetTypeOfNode`. 

PS: as I look at our type interface, my head is starting to spin, so no surprise that you're confused about which methods to call.

uditagar replied:
the `type` here is an atomic type. So I just want to check if it is a subtype of the actual node type. thanks

roshandathathri replied:
This seems wrong. It goes back to my documentation question.

I think we want the projected nodes to have all the given types. If so, then get the entity type for the given set of strings/names. Then select nodes with that entity type (no for-loop in the operator).

uditagar replied:
I can't find a function in the PropertyGraph.h file that takes a list of entity type names to return a entity type id. Thanks

roshandathathri replied:
Use `GetNodeEntityTypeManager` and then get the type from a set of strings from it.

amberhassaan commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,182 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+  new_to_old_nodes_mapping.allocateInterleaved(num_new_nodes);
+
+  Node node_idx{0};
+  for (Node src = 0; src < topology.num_nodes(); src++) {
```
We're gonna need a parallel implementation of this loop. See the implementation of `MakeNodeSortedTopo` to see how you can populate these mappings in parallel. 
Also, perhaps we won't need old to new mapping, but if you want it, see `MakeNodeSortedTopo`.

amberhassaan commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,182 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+
+  if (edge_properties.empty()) {
+    // set all edges incident to projected nodes
+    katana::do_all(katana::iterate(topology.all_nodes()), [&](auto src) {
```
why iterate over all nodes? You have the new nodes to iterate over, which could be a much smaller set.

amberhassaan commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,182 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+
+    katana::GAccumulator<uint32_t> accum_num_new_edges;
+
+    katana::do_all(katana::iterate(topology.all_nodes()), [&](auto src) {
```
Same. No need to iterate over all nodes.

amberhassaan reviewed (commented):
Left first set of comments. Highlevel comments:

* If do_all is going over nodes, where inside the loop, you're iterating over edges of a node, you should use `katana::steal()`. This is needed because node degrees can vary a lot. 
* The topology building function is quite long. Let's see if we can break it up into logical pieces, but even before that, let's see if we can simplify the logic of generating the topology. I'm reminded of the subgraph_extraction code in shared analytics.

uditagar committed:
fixes

uditagar commented:
> Left first set of comments. Highlevel comments:
> 
> * If do_all is going over nodes, where inside the loop, you're iterating over edges of a node, you should use `katana::steal()`. This is needed because node degrees can vary a lot.
> * The topology building function is quite long. Let's see if we can break it up into logical pieces, but even before that, let's see if we can simplify the logic of generating the topology. I'm reminded of the subgraph_extraction code in shared analytics.

I have added stealing now. It's faster than subgraph extraction code. That code currently checks all possible node pairs for edge.

arthurp commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,153 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+        old_to_new_nodes_mapping_(std::move(old_to_new_nodes_mapping)),
+        new_to_old_nodes_mapping_(std::move(new_to_old_nodes_mapping)),
+        old_to_new_edges_mapping_(std::move(old_to_new_edges_mapping)),
+        new_to_old_edges_mapping_(std::move(new_to_old_edges_mapping)) {}
```
This shouldn't be public I don't think.

arthurp commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,192 @@ katana::EdgeTypeAwareTopology::MakeFrom(
 
+std::unique_ptr<katana::ProjectedShuffleTopology>
+katana::ProjectedShuffleTopology::Make(
+    const katana::PropertyGraph* pg, const std::vector<std::string>& node_types,
```
This appears to be shuffling based on the type specifically. But the name just says that this is a shuffled topology. If the type is supposed to always be used for type sorting, this should be called `TypeSortedTopology` or something. If other shuffles are allow (like degree order or whatever), we need to provide ways to create views for any arbitrary shuffle and make sure we can cache them all safely.

uditagar replied:
this class is baically just projecting the node and edge types, so may be `ProjectedTypeToplogy` is a better name.

roshandathathri replied:
Yes, `Shuffle` is not appropriate here.

arthurp commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,153 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
   const Topo* topo_ptr_;
 };
 
+class KATANA_EXPORT ProjectedShuffleTopology : public GraphTopologyTypes {
```
What does this class do? This needs to documentation because it's name implies that it supports some kind of general shuffle operation, but it appears to be type based. Also we already have `BuildOrGetEdgeTypeAwareTopo` and other ones that talk about shuffling. So it's not clear how this is distinct from those.

uditagar replied:
this class creates a new local graph by filtering nodes and edges according to their types. I will add documentation then

roshandathathri replied:
Please add doxygen comments using `\\\`

roshandathathri replied:
Also, I think this should just be `ProjectedTopology`. The projection itself can be based on types (for now) or arbitrary Cypher queries (for later).

uditagar committed:
updates

uditagar committed:
added more fucntions

roshandathathri commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
 };
 
+/// filter nodes and edges based on their type values
+/// and creates a new local graph based on the filtered nodes and edges
```
Drop the word "local". All graphs are local. It can be confusing. Instead you can use "projected"

roshandathathri commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
 
+/// filter nodes and edges based on their type values
+/// and creates a new local graph based on the filtered nodes and edges
+/// also maintains mappings from old to new and new to old nodes and edges
```
Old and new can be confusing. How about "projected" and "original" instead?

uditagar replied:
sounds good

arekay commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+  Node original_node_id(const Node& nid) const noexcept {
+    return node_property_index(nid);
+  }
+
```
nit : Can we rename these to represent the conversion the methods are actually doing? `projected_to_original_edge`. It helps clear the intent and since both projected and original entities have the same type, it can be unclear what is being passed.

roshandathathri replied:
Agree. I also think it should be CamelCase: `ProjectedToOriginalEdge`. 

For now, let it be consistent with other classes. @amberhassaan

roshandathathri commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+    return edge_property_index(eid);
+  }
+
+  Edge projected_edge_id(const Edge& eid) const noexcept {
```
Document that the parameter must be original and for the above, the parameter must be projected.

roshandathathri commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+    return node_property_index(nid);
+  }
+
+  Node projected_node_id(const Node& nid) const noexcept {
```
Ditto

roshandathathri commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+    return old_to_new_edges_mapping_[eid];
+  }
+
+  PropertyIndex node_property_index(const Node& nid) const noexcept {
```
@amberhassaan I dislike that we're using snake_case in many function names. It breaks our convention of CamelCase. Unless it refers to a variable, I'm not in favor of using snake_case.

Udit, this is not an action item on you. It's better to be consistent with other graph classes, which this already is.

arekay commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,192 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+
+  katana::ParallelSTL::fill(
+      old_to_new_nodes_mapping.begin(), old_to_new_nodes_mapping.end(),
+      Node{0});
```
nit : Should this be an invalid value such as `-1` or `int_max`?

uditagar replied:
I assigned 0 here because I wanted to compute prefix sum, and then based on prefix sum values: assign new ids.

roshandathathri commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+  NUMAArray<Node> dests_;
+  NUMAArray<Node> old_to_new_nodes_mapping_;
+  NUMAArray<Node> new_to_old_nodes_mapping_;
+  NUMAArray<Edge> old_to_new_edges_mapping_;
```
Isn't this the size of the edges (and vertices above) in the original graph?

Do we need this? This is a trade-off. We can do binary search in `new_to_old_` instead. For this PR, this is fine. Please add a TODO to revisit this (also describe the trade-off).

roshandathathri reviewed (changes requested):
Except the types comment (the current code seems wrong), the rest are nits.

srogatch commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+  const Node* dest_data() const noexcept { return dests_.data(); }
+
+  /// Checks equality against another instance of ProjectedShuffleTopology.
+  /// WARNING: Expensive operation due to element-wise checks on large arrays
```
Can you put here the algorithmic complexity of how expensive those operations are? Big-O should be fine.

srogatch commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+  /// WARNING: Expensive operation due to element-wise checks on large arrays
+  /// @param that: ProjectedShuffleTopology instance to compare against
+  /// @returns true if topology arrays are equal
+  bool Equals(const ProjectedShuffleTopology& that) const noexcept {
```
Words `this` and `that` are quite similar, while here one is a pointer and the other is a const reference. Please, consider using `fellow` or `rhs`.

uditagar replied:
I will rename it to something else like `projected_topo_`

srogatch commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+      return false;
+    }
+
+    return adj_indices_ == that.adj_indices_ && dests_ == that.dests_;
```
The method is marked `noexcept`, but are you sure that the vector equality comparisons here are also `noexcept`? I believe they call element-wise comparisons, and an element comparison may throw an exception.

uditagar replied:
I will remove `noexcept` then

srogatch commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+  /// \returns iterable edge range for node.
+  edges_range edges(Node node) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(node <= adj_indices_.size());
+    edge_iterator e_beg{node > 0 ? adj_indices_[node - 1] : 0};
```
Comparison `x == 0` and `x != 0` should be slightly faster for the CPU than an open range comparison (like `x > 0`). I assume `Node` is unsigned.

srogatch commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+  edges_range edges(Node node) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(node <= adj_indices_.size());
+    edge_iterator e_beg{node > 0 ? adj_indices_[node - 1] : 0};
+    edge_iterator e_end{adj_indices_[node]};
```
You seem to trigger Undefined Behavior here if `node == adj_indices_.size()`. Either the `assert` above is wrong and should be `KATANA_LOG_DEBUG_ASSERT(node < adj_indices_.size());`, or the code in this line is wrong.

uditagar replied:
yeah, my mistake here

srogatch commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+    edge_iterator e_beg{node > 0 ? adj_indices_[node - 1] : 0};
+    edge_iterator e_end{adj_indices_[node]};
+
+    return MakeStandardRange(e_beg, e_end);
```
Are you sure it's `noexcept`?

srogatch commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+      return Node{0};
+    }
+
+    auto it = std::upper_bound(adj_indices_.begin(), adj_indices_.end(), eid);
```
`std::upper_bound` is quite error-prone: can you comment above what your intent here is? So that the reader can easily check for off-by-one and similar errors. If the array `adj_indices` is sorted, please, mention this here again.

amberhassaan replied:
`std::upper_bound` will give predictable results because `adj_indices` is sorted.

srogatch commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+    }
+
+    auto it = std::upper_bound(adj_indices_.begin(), adj_indices_.end(), eid);
+    KATANA_LOG_DEBUG_ASSERT(it != adj_indices_.end());
```
This also deseves a comment why there will always be an item in `adj_indices` that is greater than `eid`.

srogatch commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+    KATANA_LOG_DEBUG_ASSERT(it != adj_indices_.end());
+    KATANA_LOG_DEBUG_ASSERT(*it > eid);
+
+    auto d = std::distance(adj_indices_.begin(), it);
```
I would use simply `const size_t d = it - adj_indices.begin();` to ensure it's `O(1)` or it doesn't compile. We don't want linear complexity in the number of nodes for retrieving the source node for an edge, right?

srogatch commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+
+    auto d = std::distance(adj_indices_.begin(), it);
+    KATANA_LOG_DEBUG_ASSERT(static_cast<size_t>(d) < num_nodes());
+    KATANA_LOG_DEBUG_ASSERT(d > 0);
```
This one is too paranoid: as long as someone doesn't mess up your code, this shouldn't happen. But you can add a unit test.

srogatch commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+
+  Node edge_dest(Edge edge_id) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(edge_id < dests_.size());
+    return dests_[edge_id];
```
`operator[]` isn't `noexcept` I think. Though you check it in the line above, but there could be other reasons for `operator[]` to fail (e.g. in chunked arrays).

srogatch commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+  }
+
+  nodes_range nodes(Node begin, Node end) const noexcept {
+    return MakeStandardRange<node_iterator>(begin, end);
```
`MakeStandardRange()` from `external/katana/libgalois/include/katana/Range.h` is not `noexcept`. Please, change your method accordingly.

srogatch commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+    return MakeStandardRange<node_iterator>(begin, end);
+  }
+
+  nodes_range all_nodes() const noexcept {
```
Please, review all `noexcept` for whether what they call is also `noexcept`. I'll mostly stop commenting about `noexcept` from now on in this PR.

uditagar commented:
> Except the types comment (the current code seems wrong), the rest are nits.

oh it's a intersection, not union. oh man

srogatch commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,156 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+      NUMAArray<Node>&& new_to_old_nodes_mapping,
+      NUMAArray<Edge>&& old_to_new_edges_mapping,
+      NUMAArray<Edge>&& new_to_old_edges_mapping) noexcept
+      : adj_indices_(std::move(adj_indices)),
```
I think `std::move()` should be used when we are moving something without `&&` in its declaration. For a parameter with `&&` you should definitely use `std::forward<T>()`.

uditagar replied:
I only followed the convention used in other classes. I am sorry but I am not familiar with `std::forward`. thanks

roshandathathri replied:
Serge, in this case, `std::move` is required because we're using move constructors for those variables (their type does not support copy constructors). `std::forward<t>` can return lvalue (which is not desired in this case):
https://stackoverflow.com/questions/28828159/usage-of-stdforward-vs-stdmove

amberhassaan replied:
@srogatch : `&&` in parameter name is to signify that the object will be "consumed". `std::move()` is a better choice (even though `std::forward` will work with `&&`) because a named variable with `&&` may not be passed as `r-value` reference to the method/constructor being called because it has a name, thus making it `l-value` in certain situations. (`std::move()` guarantees that the parameter being passed to the constructor is always an `r-value` reference).

srogatch commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,192 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+  const auto& topology = pg->topology();
+  if (topology.empty()) {
+    ProjectedShuffleTopology pst;
+    return std::make_unique<ProjectedShuffleTopology>(std::move(pst));
```
Why is `std::move(pst)` needed here as a parameter to `make_unique`? It should work without any arguments, the same way the class is instantiated on the stack a line above.

uditagar replied:
fixed this now

uditagar commented:
@roshandathathri  I have another question rearding projecting nodes/edge according to their types. If a node n has 4 types, and the input node_types have only 2 type (but it is a subset of these 4 types of node n). Then do we select node n or not? Thanks

srogatch commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,192 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+
+  katana::ParallelSTL::fill(
+      old_to_new_nodes_mapping.begin(), old_to_new_nodes_mapping.end(),
+      Node{0});
```
It seems `0` is used as a sentinel node ID here, while it's also a legal node ID. Why not use `Node{-1}` for the sentinel?

uditagar replied:
I need 0 for calculating prefix sum later. thanks

srogatch commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,192 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+    // set all nodes
+    katana::do_all(katana::iterate(topology.all_nodes()), [&](auto src) {
+      bitset_nodes.set(src);
+      old_to_new_nodes_mapping[src] = 1;
```
Or was it a sentinel indeed? Is it another sentinel, or what are the meanings of `0` and `1`? Please, comment extensively in the code. You have at least wasted the computing power by first filling everything with `0`s, then with `1`s. If filling with zeros is needed for another conditional branch, please, move the filling with zeros there.

uditagar replied:
you are right, I will move filling zeros to the `else` part

srogatch commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,192 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+      for (auto type : node_entity_type_ids) {
+        if (pg->DoesNodeHaveType(src, type)) {
+          accum_num_new_nodes += 1;
+          accum_num_new_edges += std::distance(
```
Again, if you never suppose it to be `O(N)` please use `-` rather than `std::distance`.

srogatch commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,192 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+          accum_num_new_edges += std::distance(
+              topology.edges(src).begin(), topology.edges(src).end());
+          bitset_nodes.set(src);
+          old_to_new_nodes_mapping[src] = 1;
```
This was declared as having `Node` element type, but you seem to use it as a boolean. I may be very wrong, so comments would definitely help.

uditagar replied:
no, it isn't boolean. I will perform prefix sum on this array.

arekay commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -539,6 +725,19 @@ katana::PGViewCache::BuildOrGetEdgeTypeAwareTopo(
   }
 }
 
+katana::ProjectedShuffleTopology*
```
Should this be encapsulated in a shared pointer?

uditagar replied:
I just followed a similar convention as was done for other classes in this file.

arekay reviewed (approved):
LGTM. Minor nits.

srogatch commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,192 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+  // fill old to new nodes mapping
+  katana::ParallelSTL::partial_sum(
+      old_to_new_nodes_mapping.begin(), old_to_new_nodes_mapping.end(),
+      old_to_new_nodes_mapping.begin());
```
Only now do I see the point why there were 0s and 1s. A comment in the beginning will save readers quite some time.

srogatch commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,192 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+  katana::do_all(katana::iterate(topology.all_nodes()), [&](auto src) {
+    if (bitset_nodes.test(src)) {
+      old_to_new_nodes_mapping[src]--;
+      new_to_old_nodes_mapping[old_to_new_nodes_mapping[src]] = src;
```
I haven't thought deep enough, but I believe this yields concurrency issues. The first line will decrease an array value, separate in each thread, should be fine (but deserves a comment). The second line will assign to an array index that probably collides with `old_to_new_nodes_mapping[src+-k]` where `k` equals to the number of zeros before or after index `src` in the old `old_to_new_nodes_mapping` array before you ran `partial_sum()`. If I'm wrong, please, comprehensively describe how this parallel algorithm is correct and why there is no data race.

uditagar replied:
the index corresponding to src will only be updated in one thread, in which it is also used. I don't believe there is any concurrency issue here, since no two threads will either update/read the same location.

srogatch commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,192 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+          for (Edge e : topology.edges(old_src)) {
+            auto dest = topology.edge_dest(e);
+            if (bitset_nodes.test(dest)) {
+              bitset_edges.set(e);
```
Is `DynamicBitset` thread-safe here? Please, comment in the code.

uditagar replied:
those two bitsets are different: one corresponds to nodes and other to edges. And this one: `bitset_edges.set(e); ` is only set by a single thread for a given edge e. thanks

roshandathathri replied:
Serge, `DynamicBitset` is thread-safe on both CPU and GPU. It supports concurrent reads (test) and concurrent writes (set/reset) but not concurrent reads+writes. This should be documented in the API. If it's not clear, please feel free to edit it and let me know.

srogatch commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,192 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+  NUMAArray<Edge> out_indices;
+  out_indices.allocateInterleaved(num_new_nodes);
+
+  katana::ParallelSTL::fill(out_indices.begin(), out_indices.end(), Edge{0});
```
Please, say in a code comment what you will be doing with this.

srogatch commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,192 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+            out_dests[e_new] = dest;
+
+            old_to_new_edges_mapping[e] = e_new;
+            new_to_old_edges_mapping[e_new] = e;
```
Please, comment here about the invariants that help you avoid a race condition here.

uditagar replied:
again there is no race condition here. locations that are updated are only written by one specific thread.

srogatch reviewed (changes requested):
The off-by-one bug is definitely worth fixing.
For the rest, I can best describe my impression with a joke that I still don't understand what I have reviewed. Please, add comments and an overview of the algorithms in the code.

roshandathathri commented:
> @roshandathathri I have another question rearding projecting nodes/edge according to their types. If a node n has 4 types, and the input node_types have only 2 type (but it is a subset of these 4 types of node n). Then do we select node n or not? Thanks

Yes, it will be selected. This will be handled transparently by `graph.DoesNodeHaveType()`.

uditagar committed:
fixes

uditagar commented:
I have fixed most of the things as suggested by all people. Regarding `noexcept`, as per the slack discussion, I have decided to follow the convention used in other classes in this file. 
@serge @roshandathathri will need to check this to clear the bottleneck here. thank you.

roshandathathri commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -40,6 +41,7 @@ struct KATANA_EXPORT GraphTopologyTypes {
 
 class KATANA_EXPORT EdgeShuffleTopology;
 class KATANA_EXPORT EdgeTypeAwareTopology;
+class KATANA_EXPORT ProjectedSTypeTopology;
```
`S` seems like a typo. If it isn't, why is forward declaration required?

uditagar replied:
yes, it's a bug.

roshandathathri commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,174 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+  /// WARNING: Expensive operation due to element-wise checks on large arrays
+  /// @param that: ProjectedTopology instance to compare against
+  /// @returns true if topology arrays are equal
+  /// should take O(|V| + |E|) in the worst case
```
nit: also it's single threaded, right?

uditagar replied:
yes

roshandathathri commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,174 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+
+  /// this function creates a topology by filtering nodes and edges
+  /// @param node_types nodes with types in node_types are filtered
+  /// @param edge_types edges with types in edge_types are filtered
```
nit: filtered can mean selected or dropped. Please use "selected" instead.

roshandathathri commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,174 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
+  }
+
+  /// this function creates a topology by filtering nodes and edges
+  /// @param node_types nodes with types in node_types are filtered
```
This seems like a union, not intersection. We are doing intersection (and that's correct).
How about this?
`  /// @param node_types the types that the selected nodes must have`

roshandathathri commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -605,6 +607,174 @@ class KATANA_EXPORT BasicTopologyWrapper : public GraphTopologyTypes {
   const Topo* topo_ptr_;
 };
 
+/// filter nodes and edges based on their type values
```
Drop "based on their type values" because that's described in the `Make*` function and this class is meant to be more general.

roshandathathri commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,198 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+    auto node_entity_type_manager = pg->GetNodeTypeManager();
+
+    auto res_node_type =
+        node_entity_type_manager.GetNonAtomicEntityTypeFromStrings(node_types);
```
There's a bug here (ditto for edges): 

There is no guarantee that the vector of node types has a corresponding entity type ID. 
(1) Consider "Person" and "Message". No node has both those types, so there will no entity type ID. 
(2) Consider "XYZ". No node may have this type, so there will be no entity type ID. 

This (caller) code should check if an error is returned. An error means that no node has this type, so you can skip the code below and create a graph with no nodes and edges (if an error is returned for edge types below, create a graph with only the selected nodes and no edges).

roshandathathri reviewed (approved):
LGTM

Approving it because I don't need to review this again. Please fix the bug I mentioned before you merge.

amberhassaan commented on libgalois/include/katana/GraphTopology.h:
```diff
@@ -1032,6 +1202,7 @@ using PGViewNodesSortedByDegreeEdgesSortedByDestID =
 using PGViewBiDirectional = BasicPropGraphViewWrapper<SimpleBiDirTopology>;
 using PGViewEdgeTypeAwareBiDir =
     BasicPropGraphViewWrapper<EdgeTypeAwareBiDirTopology>;
+using PGViewProjectedGraph = BasicPropGraphViewWrapper<ProjectedTopology>;
```
@uditagar : you'll need to introduce a `ProjectedTopologyWrapper` which can be simply aliased to `BasicTopologyWrapper<ProjectedTopology>`. Reason being, `BasicPropGraphViewWrapper` takes topology wrappers by copy and `ProjectedTopology` cannot (and should not) be copied. `BasicTopologyWrapper` takes a topology object type by pointer and can be copied trivially.

uditagar replied:
I get what you are saying. So should I need to do this `using PGViewProjectedGraph = BasicTopologyWrapper<ProjectedTopology>` or do I need to create a new class. And if so, what should be the contents of this new class? Thanks

uditagar replied:
added new wrapper now

amberhassaan commented on libgalois/src/GraphTopology.cpp:
```diff
@@ -435,6 +435,198 @@ katana::EdgeTypeAwareTopology::MakeFrom(
+  const auto& topology = pg->topology();
+  if (topology.empty()) {
+    ProjectedTopology pst;
+    return std::make_unique<ProjectedTopology>(std::move(pst));
```
Why use 2 lines to do what can be done in one line?

uditagar replied:
fixed now

amberhassaan reviewed (commented):
2nd set of comments. More to come.

srogatch reviewed (approved):
Should be fine now.

uditagar committed:
fixes; added unit-test etc.

uditagar committed:
updated cmake

uditagar commented:
So I have updated the code according to the reviews by all people. This is a list of new updates:
1. Updated node and edge projection logic: `OR` for nodes and edges; and edge is picked only if both it's endoints are also picked
2. Added a new wrapper class for `ProjectedTopology`
3. Added unit-test. For ldbc with command: `./unit-test-projection ~/udit_katana_Mar30/build/inputs/current/propertygraphs/ldbc_003/ City,Comment,Company,Continent,Country,Forum HAS_CREATOR,HAS_INTEREST,HAS_MEMBER,HAS_MODERATOR,HAS_TAG,HAS_TYPE,IS_PART_OF,IS_SUBCLASS_OF,KNOWS,LIKES`, the output is `Original Num Nodes: 29946
 Original Num Edges: 43072
 New Num Nodes: 3813
 New Num Edges: 1454`
4. Added functionality to return empty graph: returns 0 node and edge graph if no nodes are selected; and returns a graph with num_new_nodes and 0 edges (if num_new_nodes are selected and no edges)

@roshandathathri @amberhassaan Please review changes, if you want to. Otherwise, I will merge it on Monday and go ahead with my distributed projection PR. Thank you.

uditagar committed:
udpated cmake
